lib/SILOptimizer/Mandatory/OSLogOptimization.cpp

```diff
@@ -0,0 +1,631 @@
+  /// \param symVal symbolic value for which SIL code needs to be emitted.
+  /// \param expectedType the expected type of the instruction that would be
+  /// computing the symbolic value `symVal`. The type is accepted as a
+  /// parameter as some symbolic values like integer constants can inhabit more
+  /// than one type.
+  /// \param builder SILBuilder that provides the context for emitting the code
+  /// for the symbolic value
+  /// \param loc SILLocation to use in the emitted instructions.
+  SILValue emitCodeForSymbolicValue(SymbolicValue symVal, SILType &expectedType,
+                                    SILBuilder &builder, SILLocation &loc);
+
+  /// Detect and emit diagnostics for errors found during evaluation. Errors
+  /// can happen due to incorrect implementation of the os log API in the
+  /// overlay or due to incorrect use of the os log API.
+  /// TODO: some of the checks here would be made redundant by a dedicated
+  /// diagnostics check that will happen before the optimization starts.
+  bool detectAndDiagnoseErrors(Optional<SymbolicValue> errorInfo,
+                               SingleValueInstruction *osLogMessageAddr,
+                               ConstExprStepEvaluator &constantEvaluator);
+
+  /// The entry point to the transformation.
+  void run() override;
+
+  // Cache a couple of SIL-level, string-related information while traversing
+  // the function body. This information is needed for constant folding string
+  // literals and recording it will avoid reconstructing it.
+  SILFunction *stringInitIntrinsic = nullptr;
+  Optional<SILType> stringMetatype = None;
+
+  /// Extract and cache the required string-related information from the
+  /// given instruction, if possible.
+  void extractStringInfoFromInstruction(SILInstruction *inst);
+};
+
+/// Find the SILBasicBlock::iterator for a given instruction.
+static SILBasicBlock::iterator
+findIteratorForInstruction(SILInstruction *keyInst) {
+  auto *bb = keyInst->getParent();
+  for (auto currI = bb->begin(), endI = bb->end(); currI != endI; currI++) {
+    if (keyInst == &(*currI)) {
+      return currI;
+    }
+  }
+  llvm_unreachable("Parent block doesn't contain instruction.");
+}
+
+/// Find an argument of type OSLogMessage from the given os log call.
+static SILValue findOSLogMessageArgument(ApplyInst *oslogCall) {
+  ASTContext &astContext = oslogCall->getFunction()->getASTContext();
+
+  for (auto argument : oslogCall->getArguments()) {
+    SILType argumentType = argument->getType();
+    auto *structDecl = argumentType.getStructOrBoundGenericStruct();
+    if (!structDecl) {
+      continue;
+    }
+
+    if (structDecl->getName() == astContext.Id_OSLogMessage) {
+      return argument;
+    }
+  }
+  llvm_unreachable("No argument of type OSLogMessage in os log call");
+}
+
+/// A utility function for inlining the apply instruction into the caller.
+/// TODO: should this function be moved to the SILInliner utility?
```

Comment:
Yes, that would make sense. AFAIK, the code is the same in the MandatoryInliner and in the PerformanceInliner

Reply:
@eeckstein I created a new [PR](https://github.com/apple/swift/pull/24488) to move this to the SILInliner and modified PerformanceInliner to use this utility.